app/common/src/options.ts

```diff
@@ -0,0 +1,194 @@
+      const key = prefix ? `${prefix}.${k}` : k
+      if (v != null && typeof v === 'object' && !Array.isArray(v)) {
+        Object.assign(out, flattenObject(v, key))
+      } else {
+        out[key] = v
+      }
+    }
+  }
+  return out
+}
+
+/** Turn a flat dotted-record into a nested object. */
+export function unflattenObject<T extends object>(flat: Flat<T>): T {
+  const out: any = {}
+  for (const [k, v] of Object.entries(flat)) {
+    const pathSegments = k.split('.')
+    let cur = out
+    for (const segment of pathSegments.slice(0, -1)) {
+      cur[segment] ??= {}
+      cur = cur[segment]
+    }
+    const last = pathSegments[pathSegments.length - 1]!
+    cur[last] = v
+  }
+  return out as T
+}
+
+/** Build URLSearchParams for non-default pass-to-web options. */
+export function buildWebAppURLSearchParamsFromArgs(args: Options): URLSearchParams {
+  const params = new URLSearchParams()
+  const entries = collectWebAppOptionsFromArgs(args)
+  Object.entries(entries).forEach(([key, val]) => params.append(key, String(val)))
+  return params
+}
+
+/** Collect non-default values of pass-to-web options from parsed args. */
+function collectWebAppOptionsFromArgs(options: Options): Record<string, string | number | boolean> {
+  const result: Record<string, string | number | boolean> = {}
+  const defaults = defaultOptions()
+  const flatDefaults = flattenObject(defaults)
+  const flatOptions = flattenObject(options)
+  for (const key of PASS_TO_WEB) {
+    const value = flatOptions[key]
+    if (value == null)
+      throw new Error(`Option ${key} not found, but needs to be passed to web application`)
+    const def = flatDefaults[key]
+    if (value != null && JSON.stringify(value) !== JSON.stringify(def)) {
+      result[key] = value
+    }
+  }
+  return result
+}
+
+/** Parse pass-to-web options from URLSearchParams. For missing values, defaults are used. */
+export function parseWebAppOptionsFromSearchParams(params: URLSearchParams): Options {
+  const out = flattenObject(defaultOptions())
+  for (const key of PASS_TO_WEB) {
+    const value = params.get(key)
+    if (value == null) continue
+    const def = out[key]
+    switch (typeof def) {
+      case 'boolean': {
+        const b = coerceBoolean(value)
+        if (b != null) {
+          out[key] = b
+        } else {
+          console.warn(`Invalid boolean value for option ${key}: ${value}, using default ${def}.`)
+        }
+        break
+      }
+      case 'number': {
+        const n = Number(value)
+        if (!Number.isNaN(n)) {
+          out[key] = n
+        } else {
+          console.warn(`Invalid number value for option ${key}: ${value}, using default ${def}.`)
+        }
+        break
+      }
+      case 'string': {
+        out[key] = value
+        break
+      }
+      default:
+        throw new Error(`Invalid option type for option ${key}: ${typeof def}.`)
+    }
+  }
+  return unflattenObject<Options>(out)
+}
```

Comment:
Side note: it would be nice to have these functions unit-tested, but I see there is no vite configuration in this package, so it's not for this PR.